### PR TITLE
Major changes

### PR DESCRIPTION
**Pull request by RasmusOrsoe**
This contains main functionality additions.

- Trainer class (v1.0)
- Predictor class (v0.1)
- loss functions 
- dynedge model 
- custom learning rate schedule
- adhoc utility 
- minor changes to sqlite_dataset.py

All of which are tested (and it works!) in `examples/`

We now have train/test functionality :-)

### CODE REVIEW

File: src/gnn_reco/components/loss_functions.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
+import torch
+import numpy as np
+from torch.nn import CrossEntropyLoss
+
+
+def LogCosh(prediction, graph, target):
```
Comment (asogaard) [inline src/gnn_reco/components/loss_functions.py:6]: As this is used as a function (compared to e.g. `CrossEntropyLoss` from `torch.nn`, which is a class) I would propose using "snake_case" here, as well as in L9 and L14 of this file. Alternatively, these three could be changed to classes to conform with torch standards. I would probably vote for the latter, but have no strong opinions as long as we're just following standard naming conventions.

File: src/gnn_reco/components/loss_functions.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    return torch.sum(torch.log(torch.cosh(((prediction[:,0]-graph[target])))))
+
+def CustomCrossEntropyLoss(prediction, graph, target):
```
Comment (asogaard) [inline src/gnn_reco/components/loss_functions.py:9]: This function, and the one above, are currently not being used — but I think that's fine, if we foresee using them. This function, however, seems to be a minimal wrapper around `torch.nn.CrossEntropyLoss`, where the `graph[target].long()` bit easily could (should?) be moved to the relevant part of the model code.

File: src/gnn_reco/components/loss_functions.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    f = CrossEntropyLoss()
+    return f(prediction,graph[target].long())
+
+
+def VonMisesSineCosineLoss(prediction, graph, target):
```
Comment (asogaard) [inline src/gnn_reco/components/loss_functions.py:14]: This function seems to make non-trivial assumptions about the structure of `prediction`. It might be beneficial to have a docstring that detail these assumptions?

File: src/gnn_reco/components/utils.py [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+from torch_geometric.data.batch import Batch
+import os
+from sklearn.preprocessing import RobustScaler
+import sqlite3
+import pandas as pd
+import pickle
+import numpy as np
+import os
+
+class EarlyStopping(object):
```
Comment (asogaard) [inline src/gnn_reco/components/utils.py:19]: From here and to L155, most of this code could probably be imported from `pytorch-ligtning` (see e.g. https://pytorch-lightning.readthedocs.io/en/1.4.0/api/pytorch_lightning.callbacks.early_stopping.html; without me knowing the details of that package) which could reduce maintenance overhead. I think we should leave the code in this PR as-is, but am mainly commenting this to hear your thoughts on this, @RasmusOrsoe and @mhaminh.

File: src/gnn_reco/components/utils.py [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+                batch_of_graphs.to(self.device)
+                target.extend(batch_of_graphs[self.target].detach().cpu().numpy())
+                predictions.extend(self.model(batch_of_graphs).detach().cpu().numpy())
+                event_nos.extend(batch_of_graphs['event_no'].detach().cpu().numpy())
+        out = pd.DataFrame(data = predictions, columns = self.output_column_names)
+        out['event_no'] = event_nos
+        out[self.target] = target
+        return out
+
+def vonmises_sine_cosine_post_processing(predictions,target):
```
Comment (asogaard) [inline src/gnn_reco/components/utils.py:189]: A similar calculation is done three places in this PR. I think we should leave the code as-is for now, but this might be one area where DRY (don't repeat yourself) could be applied later.

File: src/gnn_reco/components/utils.py [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+    predictions[target + '_pred'] = np.arctan2(predictions['sine'], predictions['cosine'])
+    predictions['k'] = abs(predictions['k'])
+    predictions['sigma'] = 1/predictions['k']
+    return predictions        
+
+
+def make_train_validation_dataloader(db, selection, pulsemap, batch_size, FEATURES, TRUTH, num_workers):
+    training_selection, validation_selection = train_test_split(selection, test_size=0.33, random_state=42)
```
Comment (asogaard) [inline src/gnn_reco/components/utils.py:197]: Not a change request, just a note: We should decide where in the pipeline, from I3 files to model inference, we want to define train-and-test-splits such that these become unambiguous and transferrable between models/runs/projects. I have always used modulo of event numbers as a reproducible way of making these splits, but am interested to hear your thoughts, @RasmusOrsoe, @mhaminh, and @philippeller.

File: src/gnn_reco/components/utils.py [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+    #         scaler = RobustScaler()
+    #         if truth == 'energy':
+    #             truth_scalers[truth] = scaler.fit(np.array(np.log10(truth_data[truth])).reshape(-1,1))
+    #         else:
+    #             truth_scalers[truth] = scaler.fit(np.array(truth_data[truth]).reshape(-1,1))
+
+    #     comb_scalers = {'truth': truth_scalers, 'input': feature_scaler}
+    #     os.makedirs(outdir + '/meta', exist_ok= True)
+    #     with open(outdir + '/meta/transformersv2.pkl','wb') as handle:
+    #         pickle.dump(comb_scalers,handle,protocol = pickle.HIGHEST_PROTOCOL)
```
Comment (asogaard) [inline src/gnn_reco/components/utils.py:264]: Should this code be commented out?

File: src/gnn_reco/data/sqlite_dataset.py [hunk 8]
```diff
@@ -116,20 +119,20 @@ def _create_graph(self, features, truth):
...
+            edge_index= None
         )
-        
+        graph.n_pulses = n_pulses
+
+        for label in labels_dict.keys():
+            graph[label] = labels_dict[label]    
+
+        for truth_key in ['energy', 'position_x', 'position_y', 'position_z', 'azimuth', 'zenith']:
+            graph[truth_key] = torch.tensor(truth_dict[truth_key], dtype = torch.float64)
```
Comment (asogaard) [inline src/gnn_reco/data/sqlite_dataset.py:134]: Am happy with all of these changes, but out of curiosity: If these are the reconstruction targets, what is the argument for treating these as custom labels rather than a standard target (`y=...`)? Is it to force named access to the targets, as opposed to integer indexing a generic target tensor?

File: src/gnn_reco/models/dynedge.py [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+from torch_scatter import scatter_min
+from torch_scatter import scatter_max
+from torch_cluster import knn_graph
+from torch_geometric.nn import EdgeConv
+from torch_geometric.nn.norm import LayerNorm
+from gnn_reco.models.utils import calculate_xyzt_homophily
+
+
+
+class dynedge_energy_xfeats(torch.nn.Module):
```
Comment (asogaard) [inline src/gnn_reco/models/dynedge.py:15]: I would suggest using "PascalCase" for class names.

File: src/gnn_reco/models/dynedge.py [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+        x2 = self.relu(x2)
+        x2 = self.nn3(x2)
+        
+        x2 = self.relu(x2)
+        x2 = self.nn4(x2)
+
+        return x2
+
+
+class dynedge_angle_xfeats(torch.nn.Module):
```
Comment (asogaard) [inline src/gnn_reco/models/dynedge.py:92]: I would suggest using "PascalCase" for class names.

File: src/gnn_reco/models/convnet.py [hunk 11]
```diff
@@ -0,0 +1,125 @@
+import torch
```
Comment (asogaard) [inline src/gnn_reco/models/convnet.py:1]: As mentioned in the general comments: As the focus is on integrating model training code in the repo quickly, I would review the model-specific files as they already work.

File: src/gnn_reco/models/dynedge.py [hunk 12]
```diff
@@ -0,0 +1,181 @@
+from math import atan2
```
Comment (asogaard) [inline src/gnn_reco/models/dynedge.py:1]: As mentioned in the general comments: As the focus is on integrating model training code in the repo quickly, I would review the model-specific files as they already work.

File: src/gnn_reco/models/utils.py [hunk 13]
```diff
@@ -0,0 +1,7 @@
+from torch_geometric.utils.homophily import homophily
+def calculate_xyzt_homophily(x,edge_index, batch):
```
Comment (asogaard) [inline src/gnn_reco/models/utils.py:2]: To an "outsider," at least, this is not a standard approach, and so a docstring detailing the intention behind this function would be useful.

**Review by asogaard — CHANGES_REQUESTED** [review]
Hi @RasmusOrsoe,
Please see my review comments. Once you have had a change to go over them, I will approve the merge. 👍

File: src/gnn_reco/models/utils.py (hunk 13 above)
Comment (RasmusOrsoe) [reply src/gnn_reco/models/utils.py:2]: Thanks. Added docstring.

File: src/gnn_reco/models/dynedge.py (hunk 12 above)
Comment (RasmusOrsoe) [reply src/gnn_reco/models/dynedge.py:1]: OK.

File: src/gnn_reco/models/convnet.py (hunk 11 above)
Comment (RasmusOrsoe) [reply src/gnn_reco/models/convnet.py:1]: OK.

File: src/gnn_reco/models/dynedge.py (hunk 10 above)
Comment (RasmusOrsoe) [reply src/gnn_reco/models/dynedge.py:92]: New commit - now just a single class

File: src/gnn_reco/components/utils.py (hunk 5 above)
Comment (RasmusOrsoe) [reply src/gnn_reco/components/utils.py:189]: I agree. I removed this as its unused.

File: src/gnn_reco/components/loss_functions.py (hunk 3 above)
Comment (RasmusOrsoe) [reply src/gnn_reco/components/loss_functions.py:14]: doc string added

File: src/gnn_reco/components/loss_functions.py (hunk 1 above)
Comment (RasmusOrsoe) [reply src/gnn_reco/components/loss_functions.py:6]: I changed the names

File: src/gnn_reco/data/i3extractor.py [hunk 14]
```diff
@@ -74,6 +75,13 @@ def __call__(self, frame) -> dict:
             'width' : [],
             'pmt_area': [],
             'rde': [],
+            'string': [],
+            'pmt_number': [],
+            'dom_number': [],
+            'pmt_dir_x': [],
+            'pmt_dir_y': [],
+            'pmt_dir_z': [],
+            'dom_type': [],
```
Comment (asogaard) [inline src/gnn_reco/data/i3extractor.py:84]: @RasmusOrsoe here, and below (L97-103) an Upgrade-specific change seems to have snuck in. Is this intentional? Because I suspect this might give errors when running the existing example code/using the default GCD files. Would it be okay to remove these changes for this PR, and include them as part of a separate, Upgrade-specific PR?
